### PR TITLE
[Mellanox] update asic and module temperature in a thread for CMIS management

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
#### Why I did it

When module is totally under software control, driver cannot get module temperature/temperature threshold from firmware. In this case, sonic needs to get temperature/temperature threshold from EEPROM. In this PR, a thread thermal updater is created to update module temperature/temperature threshold while software control is enabled.

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it

Query ASIC temperature from SDK sysfs and update hw-management-tc periodically
Query Module temperature from EEPROM and update hw-management-tc periodically

#### How to verify it

Manual test

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Review by keboliu — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@prgeor kindly reminder to review this PR

**Comment by xincunli-sonic** [comment]
Can we add a HLD to describe how would this thread work?

**Comment by prgeor** [comment]
@Junchao-Mellanox Module temperature is already read by Xcvrd DOM thread so why not use that?

image

**Comment by Junchao-Mellanox** [comment]
> @Junchao-Mellanox Module temperature is already read by Xcvrd DOM thread so why not use that?
> 
> image

Hi Prince, thanks for the comments. DOM info is updated every 1 minutes. However, this get_temperature function might be called at any time. So, we decided to have implementation not from DB.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py [hunk 1]
```diff
@@ -785,6 +791,77 @@ def get_tx_fault(self):
         api = self.get_xcvr_api()
         return [False] * api.NUM_CHANNELS if api else None
 
+    def get_temperature(self):
```
Comment (prgeor) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:794]: @Junchao-Mellanox why we need this thread to read the optics temperature when the same is available in TRANSCEIVER_DOM_SENSOR table

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py [hunk 2]
```diff
@@ -785,6 +791,77 @@ def get_tx_fault(self):
...
+            return thresh[consts.TEMP_HIGH_ALARM_FIELD]
+        return SFP_DEFAULT_TEMP_CRITICAL_THRESHOLD
+
+    def _get_temperature_threshold(self):
+        self.reinit()
+        api = self.get_xcvr_api()
+        if not api:
+            return None
+
+        thresh_support = api.get_transceiver_thresholds_support()
```
Comment (prgeor) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:857]: @Junchao-Mellanox thresholds are available in TRANSCEIVER_DOM_THRESHOLD table why read again?

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py [hunk 3]
```diff
@@ -785,6 +791,77 @@ def get_tx_fault(self):
         api = self.get_xcvr_api()
         return [False] * api.NUM_CHANNELS if api else None
 
+    def get_temperature(self):
+        try:
+            if not self.is_sw_control():
```
Comment (prgeor) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:796]: @Junchao-Mellanox can we make use of TRANSCIEVER_INFO table to decide sw or fw control?

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py (hunk 1 above)
Comment (Junchao-Mellanox) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:824]: Hi Prince, thanks for the comments. DOM info is updated every 1 minutes. However, this get_temperature function might be called at any time. So, we decided to have implementation not from DB.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py (hunk 2 above)
Comment (Junchao-Mellanox) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:887]: Hi Prince, thanks for the comments. DOM info is updated every 1 minutes. However, this get_temperature function might be called at any time. So, we decided to have implementation not from DB.

**Comment by Junchao-Mellanox** [comment]
Hi @prgeor, could you please review again?

**Review by prgeor — APPROVED** [review]

**Review by liat-grozovik — APPROVED** [review]

**Pull request merged**